src/main/kotlin/org/rust/ide/console/RsConsoleCodeFragmentContext.kt

```diff
@@ -10,42 +10,98 @@ import com.intellij.openapi.project.DumbService
 import com.intellij.openapi.project.Project
 import org.rust.cargo.project.model.cargoProjects
 import org.rust.lang.core.psi.*
-import org.rust.lang.core.psi.ext.descendantOfTypeStrict
+import org.rust.lang.core.psi.ext.RsNamedElement
+import org.rust.lang.core.psi.ext.block
+import org.rust.lang.core.psi.ext.childOfType
+import org.rust.lang.core.resolve.TYPES
+import org.rust.lang.core.resolve.findPrelude
+import org.rust.lang.core.resolve.processNestedScopesUpwards
 import org.rust.openapiext.toPsiFile
 
-class RsConsoleCodeFragmentContext {
+class RsConsoleCodeFragmentContext(codeFragment: RsReplCodeFragment?) {
 
-    private val topLevelElements: MutableMap<String, String> = mutableMapOf()
-    private val allCommandsText: StringBuilder = StringBuilder()
+    private val itemsNames: MutableSet<String> = mutableSetOf()
+    private val commands: MutableList<String> = mutableListOf()
 
-    fun addToContext(lastCommandContext: RsConsoleOneCommandContext) {
-        topLevelElements.putAll(lastCommandContext.topLevelElements)
-        allCommandsText.append(lastCommandContext.statementsText)
+    init {
+        if (codeFragment != null) {
+            DumbService.getInstance(codeFragment.project).runWhenSmart {
```

Comment:
Looks like the user can redefine some items from prelude before this code is invoked. I mean completion may be invalid in such case:
```Rust
>> let foo = Some(1);
>> struct Some(i32, i32);
// indexing ends, `addItemsNamesFromPrelude` is invoked
>> foo./*caret*/
```

Reply:
Thanks, fixed by forcing creating new scope if names for items from prelude are not yet added